gpg_key_integration_test.go

```diff
@@ -0,0 +1,169 @@
+	client := testClient(t)
+	ctx := context.Background()
+
+	org, orgCleanup := createOrganization(t, client)
+	t.Cleanup(orgCleanup)
+
+	upgradeOrganizationSubscription(t, client, org)
+
+	provider, providerCleanup := createRegistryProvider(t, client, org, PrivateRegistry)
+	t.Cleanup(providerCleanup)
+
+	gpgKey, gpgKeyCleanup := createGPGKey(t, client, org, provider)
+	t.Cleanup(gpgKeyCleanup)
+
+	t.Run("when the gpg key exists", func(t *testing.T) {
+		fetched, err := client.GPGKeys.Read(ctx, GPGKeyID{
+			Namespace: provider.Organization.Name,
+			KeyID:     gpgKey.KeyID,
+		})
+		require.NoError(t, err)
+
+		assert.NotEmpty(t, gpgKey.ID)
+		assert.NotEmpty(t, gpgKey.KeyID)
+		assert.Greater(t, len(gpgKey.AsciiAmor), 0)
+		assert.Equal(t, fetched.Namespace, provider.Organization.Name)
+	})
+
+	t.Run("when the key does not exist", func(t *testing.T) {
+		_, err := client.GPGKeys.Read(ctx, GPGKeyID{
+			Namespace: provider.Organization.Name,
+			KeyID:     "foobar",
+		})
+		assert.ErrorIs(t, err, ErrResourceNotFound)
+	})
+}
+
+func TestGPGKeyUpdate(t *testing.T) {
+	client := testClient(t)
+	ctx := context.Background()
+
+	org, orgCleanup := createOrganization(t, client)
+	t.Cleanup(orgCleanup)
+
+	upgradeOrganizationSubscription(t, client, org)
+
+	provider, providerCleanup := createRegistryProvider(t, client, org, PrivateRegistry)
+	t.Cleanup(providerCleanup)
+
+	// We won't use the cleanup method here as the namespace
+	// is used to identify a key and that will change due to the update
+	// call. We'll need to manually delete the key.
+	gpgKey, _ := createGPGKey(t, client, org, provider)
+
+	t.Run("when updating to a valid namespace", func(t *testing.T) {
+		// Create a new namespace to update the key with
+		org2, org2Cleanup := createOrganization(t, client)
+		t.Cleanup(org2Cleanup)
+
+		provider2, provider2Cleanup := createRegistryProvider(t, client, org2, PrivateRegistry)
+		t.Cleanup(provider2Cleanup)
+
+		keyID := GPGKeyID{
+			Namespace: provider.Organization.Name,
+			KeyID:     gpgKey.KeyID,
+		}
+		opts := GPGKeyUpdateOptions{
+			Namespace: provider2.Organization.Name,
+		}
+
+		updatedKey, err := client.GPGKeys.Update(ctx, keyID, opts)
+		require.NoError(t, err)
+
+		assert.Equal(t, gpgKey.KeyID, updatedKey.KeyID)
+		assert.Equal(t, updatedKey.Namespace, provider2.Organization.Name)
+
+		// Cleanup
+		err = client.GPGKeys.Delete(ctx, GPGKeyID{
+			Namespace: provider2.Organization.Name,
+			KeyID:     updatedKey.KeyID,
+		})
+		require.NoError(t, err)
+	})
+}
+
+func TestGPGKeyDelete(t *testing.T) {
+	client := testClient(t)
+	ctx := context.Background()
+
+	org, orgCleanup := createOrganization(t, client)
+	t.Cleanup(orgCleanup)
+
+	upgradeOrganizationSubscription(t, client, org)
+
+	provider, providerCleanup := createRegistryProvider(t, client, org, PrivateRegistry)
+	t.Cleanup(providerCleanup)
+
+	gpgKey, _ := createGPGKey(t, client, org, provider)
+
+	t.Run("when a key exists", func(t *testing.T) {
```

Comment:
What do you think about adding a test if the key doesn't exist or is invalid?

Reply:
This will return a 404, however we don't need to add a test for this. The reason being is the client in `client.do()` decodes the headers and returns the appropriate generic error. In this case that would be `ErrResourceNotFound`.